mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -973,41 +997,53 @@ struct GridwiseGemmV2RewritePattern
-        (kpack > 1) ? GemmDimension::K : GemmDimension::MorN;
-    std::tie(aVectorDim, aVectorLen) =
-        bestVectorization(b, matA, aCopyPerThread, vectorTiebreaker, kPerBlock,
-                          mPerBlock, elementType);
-    std::tie(bVectorDim, bVectorLen) =
-        bestVectorization(b, matB, bCopyPerThread, vectorTiebreaker, kPerBlock,
-                          nPerBlock, elementType);
-
-    LLVM_DEBUG(llvm::dbgs()
-               << "gridSize: " << gridSize << "\n"
-               << "blockSize: " << blockSize << "\n"
-               << "aCopyPerThread: " << aCopyPerThread << "\n"
-               << "bCopyPerThread: " << bCopyPerThread << "\n"
-               << "aVectorDim: " << aVectorDim << "\n"
-               << "aVectorLen: " << aVectorLen << "\n"
-               << "bVectorDim: " << bVectorDim << "\n"
-               << "bVectorLen: " << bVectorLen << "\n"
-               << "vectorTiebreaker: " << vectorTiebreaker << "\n");
+    // When we use XDLops we always have a [K0, M/N, K1] LDS buffer
+    // so we should always try to maximize the vectorization in K
+    // first and then try a best-effort vectorization on the other
+    // dimension
+    int64_t maxVlen = 128 / elementType.getIntOrFloatBitWidth();
+    auto maybeCopyAPerThread = computeCopyPerThreadV2(
+        maxVlen, aCopyPerThread, kPerBlock, mPerBlock, loc);
 
-    auto maybeCopyAPerThread = computeCopyPerThread(
-        aVectorDim, aVectorLen, aCopyPerThread, kPerBlock, mPerBlock, loc);
     if (failed(maybeCopyAPerThread))
       return maybeCopyAPerThread;
     int64_t aCopyKPerThread = (*maybeCopyAPerThread).first;
     int64_t copyMPerThread = (*maybeCopyAPerThread).second;
-
-    auto maybeCopyBPerThread = computeCopyPerThread(
-        bVectorDim, bVectorLen, bCopyPerThread, kPerBlock, nPerBlock, loc);
+    auto maybeCopyBPerThread = computeCopyPerThreadV2(
+        maxVlen, bCopyPerThread, kPerBlock, nPerBlock, loc);
     if (failed(maybeCopyBPerThread))
       return maybeCopyBPerThread;
     int64_t bCopyKPerThread = (*maybeCopyBPerThread).first;
     int64_t copyNPerThread = (*maybeCopyBPerThread).second;
 
-    LLVM_DEBUG(llvm::dbgs() << "kPerBlock: " << kPerBlock << "\n"
+    GemmDimension vectorTiebreaker =
+        (kpack > 1) ? GemmDimension::K : GemmDimension::MorN;
+    std::tie(aVectorDim, aVectorLen) =
+        bestVectorization(b, matA, aCopyPerThread, vectorTiebreaker, kPerBlock,
+                          mPerBlock, elementType);
+    std::tie(bVectorDim, bVectorLen) =
+        bestVectorization(b, matB, bCopyPerThread, vectorTiebreaker, kPerBlock,
+                          nPerBlock, elementType);
+
+    // If we are vectorizing along dimension M or N, we cannot vectorize
+    // more than copyMPerThread
+    if (aVectorDim == GemmDimension::MorN) {
+      aVectorLen = std::min(aVectorLen, copyMPerThread);
+    }
+    if (bVectorDim == GemmDimension::MorN) {
+      bVectorLen = std::min(bVectorLen, copyNPerThread);
+    }
+
```

Comment:
For understanding, it looks like those lines are key for this PR? 
 - Effectively, you want to limit the vectorization on M or N dimension when we are trying to make the decision to do a global -> vector load
 - The reasoning behind all this is due to you know lds layout is [K, M, kPack], kPack being the fastest changing dimension
   - Since kPack is always the fastest changing dimension, we prioritize maximize gemmK vectorization length over gemmM or gemmN

Please point out if any of my summary is wrong, just so that we are on the same page with your motivations.

------

My main question on this reasoning process is: Why would global load guard by the layout of lds store?
   - On a second thought, I think we should do global load only considering about global layout. Because the overhead of global load is a lot higher than LDS operations
   - That means that we shouldn't do global load based on vectorization length(existing logic). Rather, we should do global load based on fastest changing dimension. If global layout is [gemmM(k), gemmK(cyx)], then we should always try to both load and maximize the vectorization on gemmK dimension. 
   - Similarly, when the global layout is [k, y, x, c], again we should prioritize on gemmK dimension because the load stride between the gemmK dimension can be minimized.
   - Only when we have a layout that is [c, y, x, k] should we prioritize vectorization on gemmM/N dimension

Following this thought, I'd argue that both `aVectorDim` and `bVectorDim` should be `GemmDimension::K`, because we care about [k, c, y, x] or [k, y, x, c] layout a lot more than other layouts.

Reply:
> it feels like what the CK people found is that, since our global memory loads are software pipelined across a big block of computations, while our LDS operations aren't

Seems to be a reasonable point, the global load indeed get pipeline so penalty might be smaller

> That is, if we see "vectorization of 1, smallest stride in requested dimension = 1" and "vectorization of 1, smallest stride in requested dimension = H*W", we should pick the former.

Agreed.

Reply:
Maybe we can have a chat and talk about this? I am free in 1h30